### PR TITLE
Move bundler to Microbundle

### PR DESCRIPTION
**Pull request by Remeic**
First of all, Hi! 
This is amazing work

I submit you this PR for change the bundler.
I choose microbundle because is very easy to use and allow CodeJar to have multiple output format.

**Why?**

Some days ago, i'm starting to use CodeJar in one of my project.
When i try to add it to my unit test ( Jest + Typescript ) i have some issue due the format of the lib served by npm.
Jest can't easily import ES module ( i think this is the main problem ).

The output bundle size is the same, i have paid attention to the bundle size ( i think one of the most important feature of CodeJar)

I have choosed to serve addons Line Numbers and Cursor in specific folder.
Users can import addons as
`import {cursor} from "@medv/codejar/cursor" `

I'm not able to test if addons works properly since i think the only method is publish on npm.

What do you think about this solution? 
Can improve CodeJar experience?

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+v12.16.1
```
Comment (antonmedv) [inline .nvmrc:1]: Why to include this? There are plenty on dot files/configs. For example I use brew for node.

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,5 @@
 /node_modules/
+dist/
```
Comment (antonmedv) [inline .gitignore:2]: Is this BC?  Users expect `codejar.js` at the root.

File: cursor/package.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "cursor",
```
Comment (antonmedv) [inline cursor/package.json:2]: CodeJar is small for a resulting bundle. There is no benefit in package size. 

Why this configuration file? It adds too much bookkeeping without any benefits. 

I like what CodeJar currently requires as little as possible (only typescript).

File: package.json [hunk 4]
```diff
@@ -1,23 +1,61 @@
...
-    "*.ts",
-    "*.js"
+    "src",
+    "dist",
+    "cursor/dist",
+    "cursor/src",
+    "cursor/package.json",
+    "line_numbers/dist",
+    "line_numbers/src",
+    "line_numbers/package.json"
```
Comment (antonmedv) [inline package.json:22]: Wow, too much work.

File: package.json [hunk 5]
```diff
@@ -1,23 +1,61 @@
...
   ],
   "scripts": {
-    "start": "tsc -w",
-    "size": "minify codejar.js --sourceType module | gzip-size"
+    "build": "npm-run-all --parallel build:*",
+    "build:codejar": "microbundle --external none",
+    "build:cursor": "microbundle --cwd cursor",
+    "build:linenumbers": "microbundle --cwd line_numbers",
```
Comment (antonmedv) [inline package.json:28]: Again imagine if I'm going to add much more functionality. Before for me, it was just adding one file. Now, this is error-prone.

File: package.json [hunk 6]
```diff
@@ -1,23 +1,61 @@
...
-    "babel-minify": "^0.5.1",
     "gzip-size-cli": "^3.0.0",
+    "microbundle": "^0.12.0-next.8",
+    "npm-run-all": "^4.1.5",
     "typescript": "^3.8.3"
+  },
+  "mangle": {
+    "regex": "^_"
+  },
+  "exports": {
```
Comment (antonmedv) [inline package.json:41]: 🤯

File: src/codejar.ts [hunk 7]
```diff
@@ -13,7 +13,7 @@ type Position = {
   dir?: "->" | "<-"
 }
 
-export function CodeJar(editor: HTMLElement, highlight: (e: HTMLElement) => void, opt: Partial<Options> = {}) {
+export const CodeJar = (editor: HTMLElement, highlight: (e: HTMLElement) => void, opt: Partial<Options> = {}) => {
```
Comment (antonmedv) [inline src/codejar.ts:16]: Why this change?

**Comment by antonmedv** [comment]
Hi,

Thanks for pull requires and for warm words. It's really good to hear a positive response to your work.

Let's try to solve your problem. Or Jest problem.

**Comment by Remeic** [comment]
I'm response to every question.
I'm so sorry for some mistakes, this is one of my first serious PR.

When i submit you the PR, i know that might be an over engineering solution, but if you want we can discuss and i can try another approach.

File: .nvmrc (hunk 1 above)
Comment (Remeic) [reply .nvmrc:1]: For install microbundle next version, is needed a node version > than 10.
So i use this file for switch node version with nvm

File: .gitignore (hunk 2 above)
Comment (Remeic) [reply .gitignore:2]: BC? 
All bundled files go to dist folder, but whit package.json fields users doesn't see anything different 

`"main": "dist/codejar.js",
  "module": "dist/codejar.module.js",
  "unpkg": "dist/codejar.umd.js",`

**Comment by antonmedv** [comment]
So the main problem is what Jest doesn't work with ES modules?

Maybe new version of Jest will work? Or is it possible to configure your bundler (webpack, rollup) to process CodeJar to commonjs too?

In nothing is working, let's add commonjs.

File: cursor/package.json (hunk 3 above)
Comment (Remeic) [reply cursor/package.json:2]: This config file is needed for microbundle to bundle "sub" packages, like preact do.
So an user can import cursor by `import {cursor} from "@medv/codejar/cursor"`

Btw, i know that this solution is over engineering, microbundle doesn't support different input files yet.

File: package.json (hunk 4 above)
Comment (Remeic) [reply package.json:22]: This is need for better user suggestion

File: package.json (hunk 5 above)
Comment (Remeic) [reply package.json:28]: Yes, one more file need a change of build config

File: src/codejar.ts (hunk 7 above)
Comment (Remeic) [reply src/codejar.ts:16]: This is my fault, i commit this even if i don't want... I'm so sorry

**Comment by Remeic** [comment]
> So the main problem is what Jest doesn't work with ES modules?
> 
> Maybe new version of Jest will work? Or is it possible to configure your bundler (webpack, rollup) to process CodeJar to commonjs too?
> 
> In nothing is working, let's add commonjs.

For what i see, there is a problem with jest, but i use the latest jest version.
I think can be useful have multiple file format to serving, so i make this PR, but if the direction of codejar is serve only ES module i understand.

I'm sorry for some issue related this PR, as i said, is one of my first serious PR, but i'm thankful to you for this opportunity

File: .nvmrc (hunk 1 above)
Comment (antonmedv) [reply .nvmrc:1]: But I don't use nvm. Why include it in the repo?

File: .gitignore (hunk 2 above)
Comment (antonmedv) [reply .gitignore:2]: BC == Breaking Change.

What is users import files directly? For example via `import {} from 'https://unpkg.com/@medv/codejar@1.0.8/codejar.js'`?

**Comment by antonmedv** [comment]
Yes, my plan is to serve only modern JavaScript with CodeJar. As now a lot of tools/browsers support it out of the box. 

Maybe it's okay to have some compatibility packages (like `npm install codejar-commonjs`)

**Comment by Remeic** [comment]
@antonmedv Oh yes, this can be a good idea, how do you think this can be achieved?

Another package? Moving codejar to an organization? 

If you want, I'd like it very much help to achieve this

**Comment by antonmedv** [comment]
I think new organisation is overkill. Just a new package.

**Comment by Remeic** [comment]
> I think new organisation is overkill. Just a new package.

There something can i do for help you tho achieve this?

**Comment by antonmedv** [comment]
You can create this package on your own.

**Comment by Remeic** [comment]
> You can create this package on your own.

Ok thanks @antonmedv!
